docs/proposals/20190919-machinepool-api.md

```diff
@@ -0,0 +1,419 @@
+of machines to infrastructure provider supplied controllers.
+
+### User Stories
+
+- As an infrastructure provider author, I would like to build a controller to manage multiple
+  machines with a common configuration using my provider specific resource for doing so.
+- As a cluster operator, I would like to use MachinePool, similar to how I'm using MachineDeployment
+  today, to manage a set of machines with a common configuration.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model Changes
+
+MachinePool Spec and Status introduces the integration point for delegating the management of a set
+of machines to the infrastructure provider. Many of the fields are shared with MachineDeployment due
+to infrastructure provider's desire to enable the management of a set of machines with a single
+configuration.
+
+``` go
+type MachinePoolSpec struct
+```
+
+- **To add**
+  - **ClusterName [required]**
+    - Type: `string`
+    - Description: Name of the Cluster this machine pool belongs to.
+  - **Replicas [optional]**
+    - Type: `*int32`
+    - Description: Number of desired machine instances. Defaults to 1.
+  - **Template [required]**
+    - Type: `MachineTemplateSpec`
+    - Description: Machine Template that describes the configuration of each machine instance in a
+      machine pool.
+  - **Strategy [optional]**
+    - Type: `*MachineDeploymentStrategy`
+    - Description: Strategy to employ in replacing existing machine instances in a machine pool.
+  - **MinReadySeconds [optional]**
+    - Type: `*int32`
+    - Description: Minimum number of seconds for which a newly created machine should be ready.
+  - **ProviderIDs [optional]**
+    - Type: `[]string`
+    - Description: ProviderIDs contain a ProviderID for each machine instance that's currently
+      managed by the infrastructure provider belonging to the machine pool.
+
+``` go
+type MachinePoolStatus struct
+```
+
+- **To add**
+  - **NodeRefs [optional]**
+    - Type: `[]corev1.ObjectReference`
+    - Description: NodeRefs contain a NodeRef for each ProviderID in MachinePoolSpec.ProviderIDs.
+  - **Replicas [optional]**
+    - Type: `*int32`
+    - Description: Replicas is the most recent observed number of replicas.
+  - **ReadyReplicas [optional]**
+    - Type: `*int32`
+    - Description: The number of ready replicas for this MachinePool.
+  - **AvailableReplicas [optional]**
+    - Type: `*int32`
+    - Description: The number of available replicas (ready for at least minReadySeconds) for this
+      MachinePool.
+  - **UnavailableReplicas [optional]**
+    - Type: `*int32`
+    - Description: Total number of unavailable machines targeted by this machine pool. This is the
+      total number of machines that are still required for this machine pool to have 100% available
+      capacity. They may either be machines that are running but not yet available or machines that
+      still have not been created.
+  - **ErrorReason [optional]**
```

Comment:
For v1alpha3, we have renamed this to FailureReason. Would you please update this to match? See https://github.com/kubernetes-sigs/cluster-api/blob/7a1f750f2913002db2019fd344b54bc585087ec8/api/v1alpha3/machine_types.go#L98-L115 for the field on Machine